hashintel/hash
# Re-enable file uploads in HASH

## 🌟 What is the purpose of this PR?



This PR revives the existing file uploading mechanism in HASH using a local file uploader by default (stores in `var/uploads` in the root of the workspace).
I've made sure to test with S3 storage as well with success (see demo at the end)

Files are uploaded using the account-id joined to a generated unique identifier. This identifier is not the same as any identifier for the entity, but is treated as an external key (for file uploads, not links). It is not possible right now to "update" a file with a new version, but it is expected that the unique key will change for new editions of an entity and naturally let the file entities be immutable.

## 🔗 Related links





- [Asana task](https://app.asana.com/0/1203363157432094/1202636344465854/f) _(internal)_

## 🚫 Blocked by




- [x] #1797 

## 🔍 What does this change?




- Fixes `StorageProvider` and `UploadableStorageProvider` in the HASH API and remove some inconsistencies in the API
- Add a new `File` system type that is a thin wrapper around an external file resource
- Fix "deprecated" File GQL resolvers to match the new expected shape of the BP spec
- Add new endpoint to proxy files uploaded to HASH to be able to cache presigned URLs
- Add integration tests for files
- Add BlockProtocol hook (through shim) that should match the shape of https://blockprotocol.org/docs/spec/graph-service#message:uploadFile
- Add small demo to http://localhost:3000/type-editor (we should likely consider renaming this to some sort of demo page instead) using the BP hook to upload a file

## 📜 Does this require a change to the docs?



- New GQL resolvers have documentation

## ⚠️ Known issues


- The BP hook doesn't expose a proper loading state. As I am not using this actively anywhere, I left this as a future addition for us to do.

## 🐾 Next steps


- Implement user avatars through file uploads [Asana task](https://app.asana.com/0/1201095311341924/1201438359115690/f) (_internal_)
- Consider auth for file access
- Allow file entity "updating"

## 🛡 What tests cover this?



- Integration tests and manual tests

## ❓ How to test this?



1.  Checkout the branch / view the deployment
1.  Start the app as usual with `yarn external-services up` and `yarn dev
1.  Go to http://localhost:3000/type-editor (sign in beforehand if required)
1.  Try to upload a file and observe the result

If you wish to test the S3 backend, you must set the following values in your `.env.local` (documented in the README already)
```
FILE_UPLOAD_PROVIDER=AWS_S3
AWS_REGION=us-east-1
AWS_ACCESS_KEY_ID=CHANGEME
AWS_SECRET_ACCESS_KEY=CHANGEME
AWS_S3_UPLOADS_BUCKET=hash-file-uploads-dev
```
using your own S3 access key/secret and an existing S3 bucket you have access to

## 📹 Demo



https://user-images.githubusercontent.com/6988668/212074543-b9fa737b-1ea7-41f2-8c19-8c2c921a4750.mp4

(the image loads "slowly" as it's a 1MB file that is fetched from AWS)
